lib/screens/Badge/index.dart

```diff
@@ -0,0 +1,193 @@
+                                    fontWeight: FontWeight.w300,
+                                    fontFamily: "Roboto")),
+                          ),
+                        ]),
+                      )
+                    : new Center(
+                        child: new Column(
+                          children: <Widget>[
+                            new CircularProgressIndicator(),
+                            new Padding(
+                              padding: EdgeInsets.only(top: 10),
+                              child: new Text("Redeeming badge...",
+                                  style: new TextStyle(
+                                      fontSize: 16.0,
+                                      color: const Color(0xFF000000),
+                                      fontWeight: FontWeight.w100,
+                                      fontFamily: "Roboto")),
+                            ),
+                          ],
+                        ),
+                      ),
+              ),
+            ],
+          ),
+        ),
+      ),
+    );
+  }
+
+  void scanQRCode(int badgeId) async {
+    String link = await scanner.scan();
+
+    RegExp regExp = new RegExp(
+      ".*https:\/\/moonstone.seium.org\/user\/(([A-Za-z0-9]+-*)+)",
+      caseSensitive: false,
+      multiLine: false,
+    );
+
+    String stText;
+    Color stColor;
+
+    // check if link is valid
+    if (regExp.hasMatch(link)) {
+      List<String> vars = link.split("/");
+      String userId = vars[vars.length - 1];
+      // set the loading indicator
+      setState(() {
+        isLoading = true;
+      });
+
+      // get auth token
+      SharedPreferences prefs = await SharedPreferences.getInstance();
+      String token = prefs.getString('token');
+
+      // set request params
+      var body = {
+        "redeem": {"attendee_id": userId, "badge_id": badgeId.toString()}
+      };
+      String jsonBody = json.encode(body);
+      print(jsonBody);
+      final encoding = Encoding.getByName('utf-8');
+
+      // send request
+      final response = await http.post(
+          DotEnv().env['API_URL'] + 'api/v1/redeems',
+          headers: {
+            "Content-Type": "application/json",
+            "Authorization": "Bearer " + token
+          },
+          body: jsonBody,
+          encoding: encoding);
+
+      // change screen state regarding request result
+      if (response.statusCode == 201) {
+        stText = "Badge redeemed successfully.";
+        stColor = Colors.green;
+      } else {
+        Errors errors = new Errors.fromJson(json.decode(response.body));
+        Results res = errors.results;
+        stText = res.detail != null ? res.detail : res.msgs[0];
+        stColor = Colors.red;
+      }
+    } else {
+      stText = "Invalid link.";
+      stColor = Colors.red;
+    }
+
+    setState(() {
+      isLoading = false;
+      statusText = stText;
+      statusColor = stColor;
+    });
+  }
```

Comment:
This method is too long. Extract some methods to make maintainability easier.

Reply:
I have shortened it.